plugin/procfs/procfs.cc

```diff
@@ -0,0 +1,475 @@
+{
+  std::vector<std::string> files_found;
+
+  if (max_results <= 0 || max_depth <= 0)
+    return files_found;
+
+  std::string path;
+  std::string pattern;
+  std::string rest;
+
+  std::string::size_type first_star= query.find_first_of("*?[");
+  std::string::size_type last_sep_before_star=
+      query.find_last_of("/", first_star);
+  std::string::size_type first_sep_after_star=
+      query.find_first_of("/", first_star);
+
+  if (last_sep_before_star == std::string::npos)
+  {
+    path= std::string(".");
+    pattern= query;
+  }
+  else if (first_sep_after_star == std::string::npos)
+  {
+    path= std::string(query.begin(), query.begin() + last_sep_before_star);
+    pattern=
+        std::string(query.begin() + last_sep_before_star + 1, query.end());
+  }
+  else
+  {
+    path= std::string(query.begin(), query.begin() + last_sep_before_star);
+    pattern= std::string(query.begin() + last_sep_before_star + 1,
+                         query.begin() + first_sep_after_star);
+    rest= std::string(query.begin() + first_sep_after_star + 1, query.end());
+  }
+
+  if (rest.size() == 0)
+  {
+    return limited_glob_files(path, pattern, max_results);
+  }
+
+  DIR *          dir= opendir(path.c_str());
+  struct dirent *dir_entry;
+
+  if (!dir)
+    return files_found;
+
+  while ((dir_entry= readdir(dir)) != 0)
+  {
+    if (!fnmatch(pattern.c_str(), dir_entry->d_name,
+                 FNM_CASEFOLD | FNM_NOESCAPE | FNM_PERIOD))
+    {
+      if (dir_entry->d_type == DT_DIR || dir_entry->d_type == DT_LNK ||
+          dir_entry->d_type == DT_UNKNOWN)
+      {
+        std::string new_pattern= std::string(path)
+                                     .append("/")
+                                     .append(dir_entry->d_name)
+                                     .append("/")
+                                     .append(rest);
+        std::vector<std::string> v= limited_glob(
+            new_pattern, max_results - files_found.size(), max_depth - 1);
+        files_found.insert(files_found.end(), v.begin(), v.end());
+      }
+    }
+  }
+
+  closedir(dir);
+  return files_found;
+}
+
+static void fill_files_list()
+{
+  time_t ts= time(NULL);
+
+
+  mysql_rwlock_rdlock(&LOCK_procfs_files);
+  if (ts < procfs_plugin::files_last_updated_at + IS_PROCFS_REFRESH_SEC)
+  {
+    mysql_rwlock_unlock(&LOCK_procfs_files);
+    return;
+  }
+  mysql_rwlock_unlock(&LOCK_procfs_files);
+
+  mysql_rwlock_wrlock(&LOCK_procfs_files);
+  if (ts < procfs_plugin::files_last_updated_at + IS_PROCFS_REFRESH_SEC)
+  {
+    mysql_rwlock_unlock(&LOCK_procfs_files);
+    return;
+  }
+
+  procfs_plugin::files_last_updated_at= ts;
+
+  std::ifstream procfs_cnf("procfs.cnf");
+
+  procfs_plugin::files.clear();
+  while (procfs_cnf)
+  {
+    std::string path;
+    std::getline(procfs_cnf, path);
+    if (path.rfind("/proc", 0) != 0 && path.rfind("/sys", 0) != 0)
+      continue;
+
+    if (path.find_first_of("*?[") == std::string::npos)
```

Comment:
And what if I add "/proc/../some/sensitive/data.txt"? Realpath check seems to be missing in this case.